server/src-lib/Hasura/Server/Migrate.hs

```diff
@@ -194,6 +177,108 @@ migrateCatalog migrationTime = do
           WHERE name = $1
         ) |] (Identity schemaName) False
 
+downgradeCatalog :: forall m. (MonadIO m, MonadTx m) => DowngradeOptions -> UTCTime -> m MigrationResult
+downgradeCatalog opts time = do
+    downgradeFrom =<< getCatalogVersion
+  where
+    -- downgrades an existing catalog to the specified version
+    downgradeFrom :: T.Text -> m MigrationResult
+    downgradeFrom previousVersion
+      | previousVersion == dgoTargetVersion opts = do
+          pure MRNothingToDo
+      | previousVersion /= latestCatalogVersionString = do
+          throw400 NotSupported $
+            "In order to downgrade the catalog schema using this executable, you must be running catalog version "
+            <> latestCatalogVersionString
+            <> "."
+      | otherwise = 
+          case neededDownMigrations (dgoTargetVersion opts) of
+            Left reason -> 
+              throw400 NotSupported $
+                "This downgrade path (from "
+                  <> previousVersion <> " to " 
+                  <> dgoTargetVersion opts <> 
+                  ") is not supported, because "
+                  <> reason
+            Right path -> do
+              sequence_ path 
+              setCatalogVersion (dgoTargetVersion opts) time
+              pure (MRMigrated previousVersion)
+      
+      where
+        neededDownMigrations newVersion = 
+          downgrade previousVersion newVersion 
+            (reverse (migrations (dgoDryRun opts)))
+
+        downgrade 
+          :: T.Text
+          -> T.Text 
+          -> [(T.Text, MigrationPair m)]
+          -> Either T.Text [m ()]
+        downgrade from to = step1 where
+          step1, step2 :: [(T.Text, MigrationPair m)] -> Either T.Text [m ()]
+          step1 xs | previousVersion == from = step2 xs
+          step1 [] = Left "the starting version is unrecognized."
+          step1 ((x, _):xs)
+            | x == from = step2 xs
+            | otherwise = step1 xs
+            
+          step2 [] = Left "the target version is unrecognized."
+          step2 ((x, MigrationPair{ mpDown = Nothing }):_) = 
+            Left $ "there is no available migration back to version " <> x <> "."
+          step2 ((x, MigrationPair{ mpDown = Just y }):xs)
+            | x == to = Right [y]
+            | otherwise = (y:) <$> step2 xs
```

Comment:
To be honest, I think either approach has downsides but I chose this since it avoids multiple traversals and it's encapsulated in a `where` clause anyway, so you can choose to not understand the inner bit and just understand the purpose of the outer function. In the interest of getting this past code review, I'll just copy paste in your version.

Reply:
> In the interest of getting this past code review, I'll just copy paste in your version.

That’s not the feeling I’m trying to create! I obviously don’t want this PR to just sit here, but I also don’t think there’s any need to rush things—if it takes a couple extra days it takes a couple extra days, and I’d rather you feel good about it than feel a need to acquiesce!

I’m not terribly worried about multiple traversals over a list of length ~30, but I’d still be interested in hearing what you dislike about the approach I suggested. It’s a small thing, so I’m entirely open to being convinced otherwise. But of course, if you just don’t feel like you care strong enough one way or the other, sure, I won’t stop you from just making the change and being done with it.

Reply:
> I’d still be interested in hearing what you dislike about the approach I suggested

It's not that I don't like it. I think it has its own downsides - for example, I think using `null` like this doesn't lead to the most readable code either. You have to understand the way the traversal over the list works, whereas the other version is self-contained, with an admittedly more complex implementation of its own.

It's not that I don't like it, but that it doesn't seem better enough to warrant changing, retesting with a bunch of inputs, waiting again for CI and another review.

Reply:
Those are reasonable points. I’d be happy with the implementation as-is if you gave `step1` and `step2` more meaningful names. Maybe something like `dropUpToFrom` and `takeUpToTo`? The names “from” and “to” aren’t the most illustrative here, but I think I’d still find that clearer.

Reply:
(Also, as an aside: CI hours are cheap, and I don’t ever think “waiting for another review” should be the reason not to improve something. It’ll get merged eventually! Let’s get it right.)

Reply:
I'm not saying don't fix things/don't get it right. I'm saying we don't need to fix everything including legacy code in one PR. We can always add notes in comments for things to fix later. You say the cost of CI is cheap, but our CI takes _a long time_, and the context switch between tasks is costly. I think it's better to make some incremental improvement and document what we defer. Ultimately, this isn't my decision though.

Reply:
This is admittedly a much broader discussion than this one PR, but for context: I think we’d agree that the code is not in the state we’d like it to be, but spending time just refactoring and cleaning things up is expensive. It makes sense that some of the code is messy, as much of it was startup code. Taking shortcuts when writing startup code is to be expected.

We aren’t writing startup code anymore, though, and the policy I’ve always taken on mature codebases is this:

  1. Barring really compelling exceptional cases, we should never knowingly introduce new technical debt. Inevitably, the technical debt accumulates anyway, since sometimes decisions that seem good at the time turn out to not be good; that’s just programming. But we don’t ever do it if we can help it.

     If doing something without introducing technical debt is too much work, we cut scope. That’s the knob we are allowed to turn: we can choose between getting something done quickly or making something more featureful. But we don’t compromise on our engineering principles either way.

  2. Instead of trying to refactor things in one big burst, try as much as possible to bundle small refactors and code improvements with feature work that touches that part of the code. This “leave the campground cleaner than you found it” policy helps to eliminate existing technical debt without having to allocate time specifically for paying it back, and it essentially amortizes that cost across feature work.

     The nice thing about this approach is that time spent refactoring just naturally becomes part of your velocity. If it takes longer to implement certain features because you have to disentangle some old legacy cruft to implement it, that’s okay: that’s us paying down some decisions we made a long time ago. On the other hand, if we’re implementing something in a part of the code that is nice and clean, our estimate can be lower.

These policies keep *everyone* honest, since it means someone who’s looking at a timeline can’t put pressure on engineers to cut corners to get something done faster, and it helps them feel the value of cleaning things up. It also keeps us honest, since we are forced to be pragmatic about which things we need to refactor to get our feature done and which things we can defer until another day. In other words, we own the code and make our own decisions about our principles, and people doing higher-level project planning own prioritization of work, but we don’t have to feel like we’re in opposition; we ought to share the same common goals.

Cleaning up bits of code around the code that a feature touches is good, and I’d strongly encourage doing that. In fact, I would encourage viewing that as the *primary* mechanism for making improvements to the code quality overall. Sometimes a separate refactoring project is unavoidable, but I would prefer to avoid doing that unless we have to.

---

Finally, on the cost of CI: I agree it takes a long time. I’d like to make it take less time. Currently I think it takes more time than it needs to, but the stuff it’s doing is largely very valuable, and I’d rather it take time than just get rid of those tests. Code review doesn’t really depend on CI in any way, so we should only be paying it once per PR, not once per review cycle.

As for context-switching, yes, I agree, it’s a price. If you think there are ways to streamline that cost, I’m all ears. I personally find the asynchronous nature of GH reviews to help me feel less interrupted, but I also realize some people find just the opposite. Let me know if you’d rather just get on a call and talk about something.